core/src/block.rs

```diff
@@ -175,66 +131,104 @@ mod chained {
+                ),
+                transactions,
+                rejected_transactions,
+                event_recommendations: self.0.event_recommendations,
+            })
+        }
+
+        /// Create a new blockchain with current block as the first block.
+        pub fn chain_first(
+            self,
+            transaction_validator: &TransactionValidator,
+            wsv: &WorldStateView,
+        ) -> BlockBuilder<Chained> {
+            let (transactions, rejected_transactions) = Self::validate_transactions::<true>(
+                self.0.transactions,
+                transaction_validator,
+                wsv,
+            );
+
+            BlockBuilder(Chained {
+                header: Self::make_header(
+                    self.0.timestamp_ms,
+                    0,
+                    None,
+                    0,
+                    &transactions,
+                    &rejected_transactions,
+                    self.0.commit_topology,
+                ),
+                transactions,
+                rejected_transactions,
+                event_recommendations: self.0.event_recommendations,
+            })
         }
     }
 }
 
-mod valid {
+mod chained {
     use super::*;
 
-    /// After full validation [`ChainedBlock`] can transform into [`ValidBlock`].
-    #[derive(Debug, Clone)]
-    pub struct ValidBlock {
+    /// When [`PendingBlock`] chained with a blockchain it becomes [`ChainedBlock`]
+    #[derive(Debug, Clone, Decode, Encode)]
+    pub struct Chained {
         /// Block header
-        pub header: BlockHeader,
-        /// Array of rejected transactions.
-        pub rejected_transactions: Vec<VersionedRejectedTransaction>,
-        /// Array of all transactions in this block.
-        pub transactions: Vec<VersionedValidTransaction>,
+        pub(crate) header: BlockHeader,
+        /// List of transactions which passed validation
+        pub(crate) transactions: Vec<VersionedValidTransaction>,
+        /// List of transactions which failed validation
+        pub(crate) rejected_transactions: Vec<VersionedRejectedTransaction>,
         /// Event recommendations.
-        pub event_recommendations: Vec<Event>,
+        pub(crate) event_recommendations: Vec<Event>,
     }
 
-    impl ValidBlock {
-        /// Calculate hash of the current block.
-
-        #[inline]
-        pub fn hash(&self) -> HashOf<Self> {
-            HashOf::new(&self.header).transmute()
-        }
-
+    impl BlockBuilder<Chained> {
         /// Sign this block and get [`SignedBlock`].
         ///
         /// # Errors
+        ///
         /// Fails if signature generation fails
         pub fn sign(self, key_pair: KeyPair) -> Result<SignedBlock> {
-            let signature = SignatureOf::from_hash(key_pair, &self.hash().transmute())
-                .wrap_err(format!("Failed to sign block with hash {}", self.hash()))?;
+            let hash = HashOf::new(&self.0.header).transmute();
```

Comment:
no because then it would have to be first encoded like this:
```suggestion
            let hash = Hash::new(&self.0.header.encode()).typed_unchecked();
```